hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/realtime/AbstractRealtimeRecordReader.java

```diff
@@ -111,8 +111,8 @@ public AbstractRealtimeRecordReader(RealtimeSplit split, JobConf job) {
   }
 
   private boolean usesCustomPayload(HoodieTableMetaClient metaClient) {
-    return !(metaClient.getTableConfig().getPayloadClass().contains(HoodieAvroPayload.class.getName())
-        || metaClient.getTableConfig().getPayloadClass().contains(OverwriteWithLatestAvroPayload.class.getName()));
+    return metaClient.getTableConfig().getRecordMergeMode().equals(RecordMergeMode.CUSTOM)
+        && metaClient.getTableConfig().getRecordMergerStrategy().equals(HoodieRecordMerger.PAYLOAD_BASED_MERGER_STRATEGY_UUID);
```

Comment:
Does this mean that if the user wants to migrate from using the payload class to the merger implementation class, the merger strategy ID needs to be changed?

Reply:
Makes sense.  Let's add a Jira to update the docs update so we remember it.

Reply:
HUDI-8424 to track the follow-up.